src/useSubscription.ts

```diff
@@ -0,0 +1,100 @@
+import {
+  ApolloClient,
+  ApolloError,
+  OperationVariables,
+  SubscriptionOptions,
+} from 'apollo-client';
+import { DocumentNode } from 'graphql';
+import { useEffect, useRef, useState } from 'react';
+import isEqual from 'react-fast-compare';
+import { Omit } from './utils';
+
+import { useApolloClient } from './ApolloContext';
+
+export type OnSubscriptionData<TData> = (
+  options: OnSubscriptionDataOptions<TData>
+) => any;
+
+export interface OnSubscriptionDataOptions<TData> {
+  client: ApolloClient<any>;
+  subscriptionData: SubscriptionHookResult<TData>;
+}
+
+export interface SubscriptionHookOptions<TData, TVariables>
+  extends Omit<SubscriptionOptions<TVariables>, 'query'> {
+  shouldResubscribe?: any;
+  onSubscriptionData?: OnSubscriptionData<TData>;
+}
+
+export interface SubscriptionHookResult<TData> {
+  data?: TData;
+  error?: ApolloError;
+  loading: boolean;
+}
+
+export function useSubscription<TData = any, TVariables = OperationVariables>(
+  query: DocumentNode,
+  options?: SubscriptionHookOptions<TData, TVariables>
+): SubscriptionHookResult<TData> {
+  const prevOptions = useRef<null | SubscriptionHookOptions<TData, TVariables>>(
+    null
+  );
+  const onSubscriptionData = useRef<null | OnSubscriptionData<TData>>(null);
+  const client = useApolloClient();
+  const [result, setResult] = useState<SubscriptionHookResult<TData>>({
+    loading: true,
+  });
+
+  let shouldResubscribe;
+  if (options) {
+    onSubscriptionData.current = options.onSubscriptionData || null;
+
+    shouldResubscribe = options.shouldResubscribe;
+    if (typeof shouldResubscribe === 'function') {
+      shouldResubscribe = !!shouldResubscribe();
+    }
+  }
+
+  let inputs;
+  if (shouldResubscribe === false) {
+    // never resubscribe
+    inputs = [];
+  } else if (shouldResubscribe === undefined) {
+    inputs = [
+      query,
+      isEqual(prevOptions.current, options) ? prevOptions.current : options,
+    ];
+  }
+
+  useEffect(() => {
+    if (options) {
```

Comment:
Why is this conditional? How does this handle the case where `useSubscription` might be called multiple times where `options` is only sometimes provided?

Reply:
If shouldResubscribe is undefinded input would be equal to an array contains query and options which means useEffect only resubsribes only if one of those values has been changed, if shouldResubscribe was equal to false then input would be an empty array which means useEffect never resubscribes, other wise input would be undefined and useEffect will resubscribes with every call.

Reply:
Something else worth pointing out here is that `useEffects` requires the deps (second) parameter to always be the same size array. It will throw an error if you pass an empty array one time, and an array of length 2 another. As such I think line 61 should be `[null, null]` or something similar.

Reply:
it that case I think there would be better solution here, currently it's ugly as it is.